include/download_client.h

```diff
@@ -17,220 +17,184 @@
-	/** Indicates that an error occurred and the download
-	 *  cannot continue.
+/**
+ * @brief Download client event IDs.
+ */
+enum download_client_evt_id {
+	/** Event contains a fragment */
+	DOWNLOAD_CLIENT_EVT_FRAGMENT,
+	/** Download complete */
+	DOWNLOAD_CLIENT_EVT_DONE,
+	/**
+	 * An error has occurred during download and
+	 * the connection to the server has been lost.
+	 * - ENOTCONN: error reading from socket
+	 * - ECONNRESET: peer closed connection
+	 *
+	 * In both cases, the application should
+	 * disconnect (@ref download_client_disconnect)
+	 * and connect (@ref download_client_connect)
+	 * before reattempting the download, to
+	 * reinitialize the network socket.
 	 */
-	DOWNLOAD_CLIENT_ERROR = 0xFF
+	DOWNLOAD_CLIENT_EVT_ERROR,
 };
 
-
-/** @brief Download events. */
-enum download_client_evt {
-	/** Indicates an error during download.
-	 *  The application should disconnect and retry the operation
-	 *  when receiving this event.
-	 */
-	DOWNLOAD_CLIENT_EVT_ERROR = 0x00,
-	/** Indicates reception of a fragment during download.
-	 *  The @p fragment field of the @ref download_client object
-	 *  points to the object fragment, and the fragment size
-	 *  indicates the size of the fragment.
-	 */
-	DOWNLOAD_CLIENT_EVT_DOWNLOAD_FRAG = 0x01,
-	/** Indicates that the download is complete.
-	 */
-	DOWNLOAD_CLIENT_EVT_DOWNLOAD_DONE = 0x02,
+/**
+ * @brief Download client event.
+ */
+struct download_client_evt {
+	/** The event ID */
+	enum download_client_evt_id id;
+	union {
+		/** Error reason */
+		int error;
+		/** Fragment data */
+		struct {
+			const void *buf;
+			size_t len;
+		} fragment;
+	};
 };
 
-/** @brief Download client asynchronous event handler.
+/**
+ * @brief Download client asynchronous event handler.
  *
- * The application is notified of the status of the object download
- * through this event handler.
- * The application can use the return value to indicate if it handled
- * the event successfully or not.
- * This feedback is useful when, for example, a faulty fragment was
- * received or the application was unable to process a object fragment.
+ * The application receives events through this callback, such as
+ * the download of a fragment, the completion of the download or errors.
  *
- * @param[in] client	The client instance.
- * @param[in] event     The event.
- * @param[in] status    Event status (either 0 or an errno value).
+ * If the callback returns a non-zero value, the download will stop.
+ * To resume the download, use @ref download_client_start().
  *
- * @retval 0 If the event was handled successfully.
- *           Other values indicate that the application failed to handle
- *           the event.
+ * @param[in] event	The event.
+ *
+ * @return Zero to continue the download, non-zero otherwise.
  */
-typedef int (*download_client_event_handler_t)(struct download_client *client,
-				enum download_client_evt event, u32_t status);
+typedef int (*download_client_callback_t)(
```

Comment:
Is it needed to have this return an int? AFAICS its return value is not used.

Reply:
If non-zero is returned, the download will stop.